map/gps_track.cpp

```diff
@@ -12,193 +12,144 @@
 namespace
 {
 
-size_t const kGpsCollectionMaxItemCount = 100000;
-size_t const kGpsFileMaxItemCount = 100000;
+size_t const kMaxItemCount = 100000;
 hours const kDefaultDuration = hours(24);
 
+inline pair<size_t, size_t> UnionRanges(pair<size_t, size_t> const & a, pair<size_t, size_t> const & b)
+{
+  if (a.first == GpsTrack::kInvalidId)
+  {
+    ASSERT_EQUAL(a.second, GpsTrack::kInvalidId, ());
+    return b;
+  }
+  if (b.first == GpsTrack::kInvalidId)
+  {
+    ASSERT_EQUAL(b.second, GpsTrack::kInvalidId, ());
+    return a;
+  }
+  ASSERT_LESS_OR_EQUAL(a.first, a.second, ());
+  ASSERT_LESS_OR_EQUAL(b.first, b.second, ());
+  return make_pair(min(a.first, b.first), max(a.second, b.second));
+}
+
 } // namespace
 
 size_t const GpsTrack::kInvalidId = GpsTrackCollection::kInvalidId;
 
-GpsTrack::GpsTrack(string const & filePath)
-  : m_filePath(filePath)
-  , m_duration(kDefaultDuration)
+GpsTrack::GpsTrack(string const & filePath, size_t maxItemCount, hours duration)
+  : m_maxItemCount(maxItemCount)
+  , m_filePath(filePath)
+  , m_duration(duration)
+  , m_needClear(false)
+  , m_needSendSnapshop(false)
+  , m_threadExit(false)
+  , m_threadWakeup(false)
 {
+  ASSERT_GREATER(m_maxItemCount, 0, ());
+  ASSERT(!m_filePath.empty(), ());
+  ASSERT_GREATER(m_duration.count(), 0, ());
 }
 
-void GpsTrack::AddPoint(location::GpsTrackInfo const & info)
+GpsTrack::~GpsTrack()
 {
-  lock_guard<mutex> lg(m_guard);
-
-  LazyInitFile();
-
-  // Write point to the file.
-  // If file exception happens, then drop file.
-  if (m_file)
+  unique_lock<mutex> lg(m_threadGuard);
+  if (m_thread.joinable())
```

Comment:
What's the point to call joinable? If the thread was not detached and joined previously, it's joinable.